From 52c9d63220dd30e6dbbfdd6184395ee5943dec78 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Mon, 9 Sep 2024 12:51:07 +0530
Subject: [PATCH] wifi: ath12k: fix wrong usage of word mask subscription

The current word mask subscription is only applicable to the RX ring and
not to the monitor destination ring. Using the RX ring MSDU_END structure
in the monitor ring causes invalid data fetching. Fix this issue by
defining the non-compact TLV MSDU_END structure, which is common across all
hardware, and use it in the monitor path. Additionally, remove the
sw_frame_group_id member from the MSDU_END, and correct the MSDU_END word mask
value.

Patch-dependency: 881-wifi-ath12k-add-ethtool-support-for-rx-and-tx-stats.patch

Fixes: (Ie4f1db52c92) wifi: mac80211/ath12k: add support to ethtool stats

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_mon.c  | 21 +++++++++++----------
 drivers/net/wireless/ath/ath12k/hal.c     |  2 +-
 drivers/net/wireless/ath/ath12k/hal_rx.h  |  9 +++++++++
 drivers/net/wireless/ath/ath12k/rx_desc.h |  2 --
 4 files changed, 21 insertions(+), 13 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -1706,15 +1706,15 @@ ath12k_dp_mon_rx_parse_status_tlv(struct
 		ath12k_dp_mon_hal_rx_parse_u_sig_hdr(tlv_data, ppdu_info);
 		break;
 	case HAL_RX_MSDU_END:
-		struct rx_msdu_end_qcn9274 *msdu_end =
-					(struct rx_msdu_end_qcn9274 *)tlv_data;
+		struct hal_rx_msdu_end *msdu_end =
+				(struct hal_rx_msdu_end *)tlv_data;
 		u32 errmap = 0;
 		u32 grp_id;
 
-		info[13] = __le32_to_cpu(msdu_end->info13);
+		info[2] = __le32_to_cpu(msdu_end->info2);
 
-		ath12k_dp_mon_parse_rx_msdu_end_err(info[13], &errmap);
-		info[0] = __le32_to_cpu(msdu_end->sw_frame_group_id);
+		ath12k_dp_mon_parse_rx_msdu_end_err(info[2], &errmap);
+		info[0] = __le32_to_cpu(msdu_end->info0);
 		grp_id = u32_get_bits(info[0], RX_MSDU_END_INFO0_SW_FRAME_GRP_ID);
 		if (grp_id == RX_MSDU_END_INFO0_SW_FRAMEGROUP_UCAST_DATA ||
 		    grp_id == RX_MSDU_END_INFO0_SW_FRAMEGROUP_MCAST_DATA) {
@@ -2339,14 +2339,15 @@ static int ath12k_dp_pkt_set_pktlen(stru
 static void
 ath12k_dp_mon_parse_rx_msdu_end(u8 *tlv_data, struct hal_rx_mon_ppdu_info *ppdu_info)
 {
-	struct rx_msdu_end_qcn9274 *msdu_end =
-				(struct rx_msdu_end_qcn9274 *)tlv_data;
-	u32 info = __le32_to_cpu(msdu_end->info13);
+	struct hal_rx_msdu_end *msdu_end =
+			(struct hal_rx_msdu_end *)tlv_data;
 	u32 errmap = 0, mpdu_idx, msdu_idx, decap_format;
 
-	ath12k_dp_mon_parse_rx_msdu_end_err(info, &errmap);
+	ath12k_dp_mon_parse_rx_msdu_end_err(__le32_to_cpu(msdu_end->info2),
+					    &errmap);
 
-	decap_format = u32_get_bits(info, RX_MSDU_END_INFO11_DECAP_FORMAT);
+	decap_format = u32_get_bits(__le32_to_cpu(msdu_end->info1),
+				    RX_MSDU_END_INFO11_DECAP_FORMAT);
 
 	mpdu_idx = ppdu_info->cmn_mpdu_info.mpdu_count;
 	msdu_idx = ppdu_info->cmn_mpdu_info.mon_mpdu[mpdu_idx].msdu_count;
--- a/drivers/net/wireless/ath/ath12k/hal.c
+++ b/drivers/net/wireless/ath/ath12k/hal.c
@@ -538,7 +538,7 @@ static u8 ath12k_hal_qcn9274_rx_desc_get
 }
 
 #define MPDU_START_WMASK 0xFDC
-#define MSDU_END_WMASK 0x137C3
+#define MSDU_END_WMASK 0x137C1
 
 static inline u16 ath12k_hal_rx_mpdu_start_wmask_get(void)
 {
--- a/drivers/net/wireless/ath/ath12k/hal_rx.h
+++ b/drivers/net/wireless/ath/ath12k/hal_rx.h
@@ -539,6 +539,15 @@ struct hal_rx_mpdu_start {
 	__le32 rsvd2[16];
 } __packed;
 
+struct hal_rx_msdu_end {
+	__le32 info0;
+	__le32 rsvd0[18];
+	__le32 info1;
+	__le32 rsvd1[10];
+	__le32 info2;
+	__le32 rsvd2;
+} __packed;
+
 #define HAL_RX_PPDU_END_DURATION	GENMASK(23, 0)
 struct hal_rx_ppdu_end_duration {
 	__le32 rsvd0[9];
--- a/drivers/net/wireless/ath/ath12k/rx_desc.h
+++ b/drivers/net/wireless/ath/ath12k/rx_desc.h
@@ -771,8 +771,6 @@ enum rx_msdu_start_reception_type {
 
 struct rx_msdu_end_qcn9274 {
 	__le64 msdu_end_tag;
-	__le32 sw_frame_group_id;
-	__le32 resvd1;
 	__le16 sa_sw_peer_id;
 	__le16 info5;
 	__le16 sa_idx;
